ArmEmulator/BitArithmetic.fs

```diff
@@ -258,265 +280,109 @@ module BitArithmetic
-        | Cvc -> exeDecide v false               // execute if V = 0            
-        | Chs -> exeDecide c true                // execute if C = 1
-        | Clo -> exeDecide c false               // execute if C = 0            
-        | Cge -> exeDecide n v                   // execute if N = V            
-        | Clt -> exeDecide n (not v)             // execute if N != V            
-        | Chi ->                                 // execute if C = 1 and Z = 0
-            match c with
-            | true -> exeDecide z false
-            | false -> true
-        | Cls ->                                 // execute if C = 0 or Z = 1
-            match  c with
-            | true -> exeDecide z true
-            | false -> false
-        | Cgt ->                                 // execute if Z = 0 and N = V 
-            match z with
-            | true -> false
-            | false -> exeDecide n v
-        | Cle ->                                 // execute if Z = 1 and N != V"
-            match z with
-            | true -> exeDecide n (not v)
-            | false -> false
-        | Cnv -> true
-        | Cal -> false
-
-    /// updates N Z C flags
-    let updateNZC flgs result carry =
-        let isNeg = int32 result < 0 
-        let isZero = result = 0u
-        let convCarry = 
-            match carry with
-            | 0u -> false
-            | 1u -> true
-        match isNeg,isZero with 
-        | true,false -> Ok {flgs with N = true ; Z=false ; C=convCarry}
-        | false,true -> Ok {flgs with N = false ; Z=true ; C=convCarry}
-        | _ -> Error "can't be both negative and zero"
+
+
+    let intToBool num = 
+        match num with
+        | 1u -> true
+        | 0u -> false
+        | _ -> failwithf "Invalid integer being converted to bool"
+
+
+    // functions that do the bit manipulation
+
+    /// applies the given shift to the first argument, returning the 
+    /// required result and the carry the shift produces.
+    /// i.e returns (evaluated shift number, carry)
+    let doShift n shifter shiftVal =
+        let sVal = (int32 shiftVal) % 32  
+        match shifter with 
+        | Lsl -> n <<< sVal , (((n <<< sVal - 1) >>> 31) |> intToBool)
+        | Lsr -> n >>> sVal , (((n >>> (sVal - 1)) &&& 1u) |> intToBool)
+        | Asr -> uint32 (int32 n >>> sVal) , (((uint32 (int32 n >>> (sVal - 1))) &&& 1u) |> intToBool)
+        | Ror -> (n >>> sVal) ||| (n <<<(32- sVal)) , ((((n >>> (sVal - 1)) ||| (n <<<(31 - sVal))) &&& 1u) |> intToBool)
+
+    let doRRX n carry = 
+        let newCarry = (n &&& 1u) |> intToBool        
+        (n >>> 1) + (carry <<< 31),newCarry      
 
 
     /// executes the instruction
-    let exeInstr cpuData parseOut =
+    /// takes the current CPU data (register values, flags ect), the output from parser
+    /// and the symbol table. Results in updating registers and/or flags based on 
+    /// instruction
+    let exeInstr cpuData parseOut symTable =
```

Comment:
Please could you swap the order of the operands so that it's `exeInstr cpuData symTable parseOut`?
Also, `parseOut` shouldn't be an option or a Result here -> it's guaranteed to be a `Parse<InstDecomp>` so the full function type signature should be:
`DataPath<'a> -> SymbolTable -> Parse<InstDecomp> -> Result<DataPath<'a>,string>`